Add domain decomp

arporter opened the pull request:
This is for work done in #5. It adds an option to build the library with MPI support and generates a domain decomposition for the grid.

Rather than make MPI a mandatory dependency, I've implemented two modules, `parallel_mod.f90` and `parallel_stub_mod.f90` that provide the same interfaces. The build system creates a link to one or other of these files, depending on whether DM support is required, prior to the compilation.

arporter committed:
#5 begin to extend grid type with necessary info

arporter committed:
#5 add new parallel_mod

arporter committed:
#5 add parallel_mod to makefile

arporter committed:
#5 make init/final routines public

arporter committed:
#5 implement decompose() to include variable halo width

arporter committed:
#5 add parallel_abort()

arporter committed:
#5 fix stupid error in checking decompose() args

arporter committed:
#5 create new subdomain_type and use to store decomp

arporter committed:
#5 put back output of domain stats

arporter committed:
#5 use grid%subdomain instead of grid%simulated_region

arporter committed:
#5 mv decompose() into subdomain_mod

arporter committed:
#5 add support for non-mpi build

arporter committed:
#5 add targets to make parallel src and fix moved code

arporter committed:
#5 add nprocx/y to parallel_mod

arporter committed:
#5 make get_num_ranks public in stub lib

arporter committed:
#5 change thread tiling to use subdomain_mod::decompose()

arporter committed:
#5 change to support new stencil meta-data

arporter committed:
#5 change grid_init() to use dims of decomposed grid

arporter committed:
#5 tidy decompose() for style and unused vars

arporter committed:
#5 tidy use of parallel_mod link file

arporter committed:
#5 tidy subdomain_mod for compiler warnings

arporter committed:
#5 add parallel_common to avoid duplication

arporter committed:
#5 remove now unnecessary field_mod::tile_setup

arporter committed:
#5 rm now un-used grid_mod::compute_internal_region

arporter committed:
#7 add example mini-app

arporter committed:
#7 Add decomposition_type and change grid_init to take it as arg

arporter committed:
#7 rename gocean_init->initialise, add _finalise

arporter committed:
#5 add license and description to Makefiles

arporter committed:
#5 add noddy test target for example mini-app

arporter committed:
#5 fix typo

arporter committed:
#5 tweaks to comments

arporter committed:
#6 rm duplicated code

arporter committed:
#6 rm commented-out code from tile def

arporter commented:
@hiker - this is ready for review now if you'd be so kind. I can't explicitly request a review from you yet because you weren't a collaborator on this project - you should have an invitation for that now.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+     !! boundary points. n{x,y} != {x,y}stop - {x,y}start +1 because
+     !! {x,y}start and {x,y}stop are the coordinates of the internal
+     !! region in the global domain.
+     integer :: nx, ny
```
Wouldn't it be more descriptive to use type(region_type) :: external instead of separate {x,y}{start,stop}/n{x,y}?
This would make it obvious everywhere whether you are using the external or internal domain.

arporter replied:
That's a good idea. I've used "global" rather than "external" as it holds information on where the subdomain sits within the global one.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+                     halo_width) result(decomp)
+    !> Decompose a domain consisting of domainx x domainy points
+    !! into a 2D grid.
+    !! Returns an array of tiles describing each of the subdomains.    
```
The comment must go before the function, otherwise doxygen does not pick it up.

arporter replied:
Done.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+
+  function decompose(domainx, domainy,             &
+                     ndomains, ndomainx, ndomainy, &
+                     halo_width) result(decomp)
```
Would it potentially be useful to support different halo widths in east-west and north/south direction at least? Considering that we had a long discussion on how to specify various stencils in PSyclone ;)
Saying that, I am happy to just accept one halo size ... since this will likely cover 95% of all use cases ;)

arporter replied:
I think that could be a future addition - I can imagine adding an optional `halo_widths(4)` argument to this routine. I think the rest of the library has enough flexibility to support this feature in the future.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+    end if
+
+    if(present(halo_width))then
+       if(halo_width < 1 .and. get_num_ranks() > 1)then
```
I would imagine that there are some use cases for distributed fields without halo - e.g. if you just compute an output field from other fields (likely with halo). So it might be too restrictive to enforce those fields to have a halo.

arporter replied:
Oh yes, good point. However, this routine computes a general decomposition, independent of any particular field. This might be influenced by some of your points below though so I'll hold off for now.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+             ntiley = ntilex
+             ntilex = ierr
+          end if
+       end if
```
Impressive attention to detail :+1: I would have forced the user to supply the domain decomposition, but this is certainly useful - well done!

arporter replied:
Thanks :-) It was some code I had "prepared earlier" (do you have Blue Peter in Oz?)

hiker replied:
Nope :)

arporter replied:
Oh well. Kid's TV show where they do a craft activity and then skip to the finished article with a "here's one I prepared earlier".

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+    ! actually span a height greater or less than N. If so, we try and
+    ! reduce the height of each row by just one until we've accounted
+    ! for the <jover> extra rows.
+    !nwidth = (ntiley-2)*(idy-2) + 2*(idy-1)
```
Please remove commented out code.

arporter replied:
Well spotted - done.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+    ! If we think about the internal regions of the tiles,
+    ! then they should share the domain between them:
+    internal_width = NINT(REAL(xlen) / REAL(ntilex))
+    internal_height = NINT(REAL(ylen) / REAL(ntiley))
```
This seems to me to be way too complicated - instead of rounding just always round down. This way you have only two cases to consider (junder=0 or junder>0), which should simplify the code further down as well, and I can't see any disadvantage in this approach.

arporter replied:
Oh yes. Done.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+       junder = 0
+    end if
+    ! Ditto for x dimension
+    !nwidth = (ntilex-2)*(idx-2) + 2*(idx-1)
```
Remove commented out code

arporter replied:
Done.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+       iover  = 0
+       iunder = 0
+    end if
+
```
Ditto for rounding.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+    do jj = 1, ntiley, 1
+
+       ! Point to subdomain we are about to define
+       subdomain => decomp%subdomains(ith)
```
Remove - not used here, and it is being redefined later.

arporter replied:
True but the compiler does not know that `ntilex` is guaranteed to be > 1 (and thus that the `ji` loop will do at least one trip) and therefore warns about the assignment using the `subdomain` pointer after the loop over `ji`. This is a simple way of preventing that warning. I've extended the associated comment to explain this.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+          junder = junder - 1
+       else
+          height = internal_height
+       end if
```
With rounding down in line 147 one else branch would be unnecessary.

arporter replied:
Done.

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+          subdomain%internal%xstart = hwidth+1
+          subdomain%internal%xstop = subdomain%internal%xstart+width-1
+          subdomain%internal%nx = subdomain%internal%xstop - &
+                                  subdomain%internal%xstart + 1
```
Isn't ...%nx = width? That looks a lot simpler ;)

arporter replied:
Oh yes. Ahem. :-)

hiker commented on finite_difference/src/subdomain_mod.f90:
```diff
@@ -0,0 +1,302 @@
+          subdomain%internal%ystart = hwidth+1
+          subdomain%internal%ystop = subdomain%internal%ystart+height-1
+          subdomain%internal%ny = subdomain%internal%ystop - &
+                                  subdomain%internal%ystart + 1
```
ny = height

arporter replied:
Done.

hiker reviewed (commented):
Review of the remaining code still to be done by me.

hiker commented on finite_difference/src/grid_mod.f90:
```diff
@@ -2,6 +2,7 @@ module grid_mod
   use kind_params_mod
```
Add a doxygen comment "!>" at beginning of file, otherwise it appears no doxygen code is created.

hiker replied:
Still most modules have the module information missing (e.g. in finite_difference/src/doc/html/annotated.html), but I'll only request new files to add them (and have added appropriate comments elsewhere).

hiker commented on finite_difference/src/grid_mod.f90:
```diff
@@ -121,8 +120,8 @@ function get_tmask(self) result(tmask)
-  !> Basic constructor for the grid type. Full details
-  !! are fleshed-out by the grid_init() routine.
+  !> Basic constructor for the grid type. Full details, including domain
+  !! decomposition, are fleshed-out by the grid_init() routine.
```
While it might be unrelated to this pull request - but why a two step initialisation process? Shouldn't one constructor be enough?

arporter replied:
I think I did it this way so that you can pass the grid object down to other routines that e.g. read from file in order to set-up the T mask.

hiker replied:
Makes sense, thanks for the explanation.

hiker commented on finite_difference/src/grid_mod.f90:
```diff
@@ -194,48 +193,56 @@ end function grid_constructor
+    grid%subdomain = decomp%subdomains(myrank)
+
     ! Store the global dimensions of the grid.
     if( present(tmask) )then
```
I am not sure I understand the tmask handling here: the fact that a tmask is present is not used in the code at all, the two if and else blocks appear to be identical, except for the abort statement in case of PBC and >1 MPI rank.

I could imagine code that do not use the tmask concept at all, but still want to use an DM parallelisation, which would not be possible this way.

arporter replied:
Ah! The benefits of code review. When the code was serial I used the T-mask to set-up the size of the grid. Now however I use the domain decomposition so you are right - the if block was completely unecessary.

arporter replied:
Support for the case where we have no T mask and are running in parallel is something for a future issue...

hiker commented on finite_difference/src/argument_mod.F90:
```diff
@@ -27,14 +27,16 @@ module argument_mod
 
+! Fake array to enable us to provide stencil meta-data in the args() array
+integer, public, parameter :: stencil(0:111,0:111,0:111) = 0
+
```
This seems to be unused at this stage? I don't know how this is going to be used, but that's around 5 (to 10 in case of 8 byte default integer) :)

arporter replied:
It's needed in order to make any stencil meta-data in a kernel valid Fortran. I too am concerned about the size but hopefully the compiler will optimise it out as the quantities concerned are never actually used.

rupertford replied:
I don't think this is a reasonable way to go, particularly if we are going to support larger stencils (potentially requiring up to 0:999, 0:919, 0:999). I have a solution that uses structure constructors so does not need an array at all. It works with gfortran so I assume it is OK. I am using this in the manual ompss implementation of nemolite2d. Perhaps you could create an issue and I could add this solution in?

hiker replied:
Great :)

hiker commented on finite_difference/src/field_mod.f90:
```diff
@@ -158,7 +179,28 @@ function r2d_field_constructor(grid,    &
+!$  nthreads = omp_get_max_threads()
+    WRITE (*,"(/'Have ',I3,' OpenMP threads available.')") nthreads
+    decomp = decompose(self%internal%nx, self%internal%ny, &
+                       nthreads, ntilex, ntiley)
```
Not sure if I get this: supporting tiling on an OpenMP level seems to be a good idea, but the constructor of the grid allocated tmask and other fields without the tiling. So potentially you would (unnecessarily) access remote memory when mixing tmask (etc), and fields allocated here.
Wouldn't it make more sense to:
1) compute the thread tiling in the grid constructor (or init function)
2) Then use the the field type to create the tmask (etc) variables that are part of the grid object.
This would make this function somewhat cheaper (since all this reading of environment variable etc is done only once).

That might be a big change, and I am happy to accept the code and discuss this in a separate ticket.

arporter replied:
You  are right again - I did begin to think that it was odd that I'd (originally) chosen to set-up the thread tiling repeatedly for each new field that was created. I've created #8 for this.

hiker replied:
Imho this should be tied in with PSyclone issue 109 (https://github.com/stfc/PSyclone/issues/109) - ideally this runtime library should provide information to PSyclone on what loops to generate. Then installing the runtime library and using it with PSyclone would created either a tiled or nested loops - but that's far far out of scope of this - am happy with the ticket.

hiker commented on finite_difference/src/field_mod.f90:
```diff
@@ -158,7 +179,28 @@ function r2d_field_constructor(grid,    &
+       self%tile(itile)%whole%ystop = decomp%subdomains(itile)%ystop
+       self%tile(itile)%internal = decomp%subdomains(itile)%internal
+    end do
 
```
Apparently I can't comment on line 242 - but if we support tiling as indicated above, then the openmp loop in line 242 should use the tiling information. Plus, as far as I can tell: this would also need a change in psyclone ... I am getting the feeling I am missing something :)

arporter replied:
Oh yes, it should - done. I think PSyclone should be OK for now but might require changing if we do change the way fields have tiles.

hiker commented on finite_difference/src/parallel/parallel_stub_mod.f90:
```diff
@@ -0,0 +1,77 @@
+module parallel_mod
+  !> Stub implementation to be used in place of parallel_mod when NOT
+  !! compiling with MPI
+  use parallel_common_mod
```
Not sure about coding style: do we allow generic use? I would prefer:
use parallel_common_mod, only : rank, nranks, nprocx, nprocy

arporter replied:
Done.

hiker commented on finite_difference/src/parallel/parallel_mod.f90:
```diff
@@ -0,0 +1,97 @@
+
+  !================================================
+
+  subroutine parallel_init()
```
Docstring.

arporter replied:
Well spotted. Done - and below.

hiker commented on finite_difference/src/parallel/parallel_mod.f90:
```diff
@@ -0,0 +1,97 @@
+
+  !================================================
+
+  subroutine parallel_finalise()
```
Docstring.

hiker reviewed (changes requested):
Good work. I am not sure about a few of my comments (esp. re tiling). Also I noticed quite a few modules would not create docygen, and while I initially thought it was related to missing !> at the top of a file (there is a comment somewhere), this seems not to be the right fix in other cases.

arporter committed:
#6 add global region to subdomain_type

arporter committed:
#6 simplify load-balancing adjustment and tidy style

arporter committed:
#6 further simplification now i/jover not required

arporter committed:
#6 rm old T-mask dependent set-up of grid dimensions

arporter committed:
#6 correct first-touch in field init

arporter committed:
#6 tidy parallel routines for style

arporter committed:
#6 change example to use altered subdomain structure

arporter committed:
#6 Add doc target to makefile

arporter commented:
I've added a configuration file and build target for doxygen-generated documentation.
I think I've addressed all comments.
Ready for second review...

hiker commented on finite_difference/src/tile_mod.f90:
```diff
@@ -0,0 +1,42 @@
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+!------------------------------------------------------------------------------
+! Author: A. R. Porter, STFC Daresbury Laboratory
+
```
I'd suggest that all new files at least have the doxygen module line to have the modules listed (e.g. in finite_difference/src/doc/html/annotated.html):
```
!> Module containing the definition of the tile type
```

arporter replied:
Done.

hiker commented on finite_difference/src/dl_esm_inf.doxygen:
```diff
@@ -0,0 +1,2303 @@
+# Doxyfile 1.8.6
```
Can you rename this config file to Doxyfile or doxyfile? That's the default, allowing us to run doxygen without specifying a filename.

arporter replied:
Done.

hiker commented on finite_difference/src/parallel_common_mod.f90:
```diff
@@ -0,0 +1,79 @@
+! Author: A. R. Porter, STFC Daresbury Laboratory
+
+module parallel_common_mod
+  !> Module holding elements of the parallel infrastructure that
```
Could you move those two comments lines to before the module line, then the module will be listed in doxygen.

arporter replied:
Done.

hiker reviewed (changes requested):
A few changes only, mostly (exclusively??) related to doxygen.

arporter committed:
#6 ensure all modules have doxygen markup at the beginning

arporter committed:
#6 rename doxygen config file to standard name

arporter commented:
Doxygen does now generate the full list of modules which makes me feel good :-)

Back to @hiker for 3rd(?) review...

hiker commented:
All looking good now. My above comment that some files are missing in ... annotated.html is incorrect (it apparently lists only public types, so files without a public type declaration will not be listed there).

hiker commented:
Can you just bring it up-to-date with master (I probably merged in the wrong order.? Sorry)

arporter committed:
#6 bring up-to-date with master

arporter committed:
#6 fix example code to use GO_ suffix
